File: src/org/elixir_lang/reference/CallDefinitionClause.java
package org.elixir_lang.reference;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.psi.ResolveResult;
import com.intellij.psi.impl.source.resolve.ResolveCache;
import org.elixir_lang.psi.AtUnqualifiedNoParenthesesCall;
import org.elixir_lang.psi.call.Call;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CallDefinitionClause extends PsiReferenceBase<Call> implements PsiPolyVariantReference {
    /*
     *
     * Fields
     *
     */

    @NotNull
    public final AtUnqualifiedNoParenthesesCall moduleAttribute;

    /*
     *
     * Constructors
     *
     */

    /**
     * @param call {@code foo(arg1, ...) :: return1} in {@code @spec foo(arg1, ...) :: return1}
     * @param moduleAttribute {@code @spec foo(arg1, ...) ... return1}, so that the tree doesn't have to be walked up
     *   again if there is a {@code when}.
     */
    public CallDefinitionClause(@NotNull Call call, @NotNull AtUnqualifiedNoParenthesesCall moduleAttribute) {
        super(call, TextRange.create(0, call.getTextLength()));
        this.moduleAttribute = moduleAttribute;
    }

    /*
     *
     * Instance Methods
     *
     */

    /*
     * Public Instance Methods
     */

    /**
     * Returns the array of String, {@link PsiElement} and/or {@link LookupElement}
     * instances representing all identifiers that are visible at the location of the reference. The contents
     * of the returned array is used to build the lookup list for basic code completion. (The list
     * of visible identifiers may not be filtered by the completion prefix string - the
     * filtering is performed later by IDEA core.)
     *
     * @return the array of available identifiers.
     */
    @NotNull
    @Override
    public Object[] getVariants() {
        return new Object[0];
    }

    /**
     * Returns the results of resolving the reference.
     *
     * @param incompleteCode if true, the code in the context of which the reference is
     *                       being resolved is considered incomplete, and the method may return additional
     *                       invalid results.
     * @return the array of results for resolving the reference.
     */
    @NotNull
    @Override
    public ResolveResult[] multiResolve(boolean incompleteCode) {
        return ResolveCache
                .getInstance(this.myElement.getProject())
                .resolveWithCaching(
                        this,
                        org.elixir_lang.reference.resolver.CallDefinitionClause.INSTANCE,
                        false,
                        incompleteCode
                );
    }

    /**
     * Returns the element which is the target of the reference.
     *
     * @return the target element, or null if it was not possible to resolve the reference to a valid target.
     */
    @Nullable
    @Override
    public PsiElement resolve() {
        ResolveResult[] resolveResults = multiResolve(false);
        return resolveResults.length == 1 ? resolveResults[0].getElement() : null;
    }
}

File: src/org/elixir_lang/reference/CallDefinitionClause.kt
package org.elixir_lang.reference

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.PsiReferenceBase
import com.intellij.psi.ResolveResult
import com.intellij.psi.impl.source.resolve.ResolveCache
import org.elixir_lang.psi.call.Call
import org.elixir_lang.reference.resolver.CallDefinitionClause
import org.elixir_lang.semantic.type.definition.source.Specification

/**
 * @param call `foo(arg1, ...) :: return1` in `@spec foo(arg1, ...) :: return1`
 * @param specification `@spec foo(arg1, ...) ... return1`, so that the tree doesn't have to be walked up
 * again if there is a `when`.
 */
class CallDefinitionClause(call: Call, val specification: Specification) :
    PsiReferenceBase<Call?>(call, TextRange.create(0, call.textLength)), PsiPolyVariantReference {
    /*
     *
     * Constructors
     *
     */
    /*
     *
     * Instance Methods
     *
     */
    /*
     * Public Instance Methods
     */

    override fun getVariants(): Array<Any> = emptyArray()

    /**
     * Returns the results of resolving the reference.
     *
     * @param incompleteCode if true, the code in the context of which the reference is
     * being resolved is considered incomplete, and the method may return additional
     * invalid results.
     * @return the array of results for resolving the reference.
     */
    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
        return ResolveCache
            .getInstance(myElement!!.project)
            .resolveWithCaching(
                this,
                CallDefinitionClause,
                false,
                incompleteCode
            )
    }

    /**
     * Returns the element which is the target of the reference.
     *
     * @return the target element, or null if it was not possible to resolve the reference to a valid target.
     */
    override fun resolve(): PsiElement? {
        val resolveResults = multiResolve(false)
        return if (resolveResults.size == 1) resolveResults[0].element else null
    }
}
